core/test/mpi/distributed/vector.cpp

```diff
@@ -0,0 +1,334 @@
+                 gko::Array<nz_type>{this->ref, I<nz_type>{{0, 0, 0},
+                                                           {0, 1, 1},
+                                                           {1, 0, 2},
+                                                           {1, 1, 3},
+                                                           {2, 0, 4},
+                                                           {2, 1, 5},
+                                                           {3, 0, 6},
+                                                           {3, 1, 7},
+                                                           {4, 0, 8},
+                                                           {4, 1, 9},
+                                                           {5, 0, 10},
+                                                           {5, 1, 11}}}};
+    auto part = gko::share(
+        part_type::build_from_contiguous(this->ref, {this->ref, {0, 2, 4, 6}}));
+    auto vec = TestFixture::dist_vec_type::create(this->ref, this->comm);
+    auto rank = this->comm.rank();
+    I<I<value_type>> ref_data[3] = {
+        {{0, 1}, {2, 3}},
+        {{4, 5}, {6, 7}},
+        {{8, 9}, {10, 11}},
+    };
+
+    vec->read_distributed(md, part);
+
+    GKO_ASSERT_EQUAL_DIMENSIONS(vec->get_size(), gko::dim<2>(6, 2));
+    GKO_ASSERT_EQUAL_DIMENSIONS(vec->get_local()->get_size(),
+                                gko::dim<2>(2, 2));
+    GKO_ASSERT_MTX_NEAR(vec->get_local(), ref_data[rank], r<value_type>::value);
+}
+
+TYPED_TEST(Vector, CanReadGlobalMatrixDataScattered)
+{
+    using md_type = typename TestFixture::md_type;
+    using part_type = typename TestFixture::part_type;
+    using value_type = typename TestFixture::value_type;
+    md_type md{{0, 1}, {2, 3}, {4, 5}, {6, 7}, {8, 9}, {10, 11}};
+    auto part = gko::share(part_type::build_from_mapping(
+        this->ref, {this->ref, {0, 1, 2, 0, 2, 0}}, 3));
+    auto vec = TestFixture::dist_vec_type::create(this->ref, this->comm);
+    auto rank = this->comm.rank();
+    gko::dim<2> ref_size[3] = {{3, 2}, {1, 2}, {2, 2}};
+    I<I<value_type>> ref_data[3] = {
+        {{0, 1}, {6, 7}, {10, 11}},
+        {{2, 3}},
+        {{4, 5}, {8, 9}},
+    };
+
+    vec->read_distributed(md, part);
+
+    GKO_ASSERT_EQUAL_DIMENSIONS(vec->get_size(), gko::dim<2>(6, 2));
+    GKO_ASSERT_EQUAL_DIMENSIONS(vec->get_local()->get_size(), ref_size[rank]);
+    GKO_ASSERT_MTX_NEAR(vec->get_local(), ref_data[rank], r<value_type>::value);
+}
+
+
+TYPED_TEST(Vector, CanReadLocalMatrixData)
+{
+    using md_type = typename TestFixture::md_type;
+    using part_type = typename TestFixture::part_type;
+    using value_type = typename TestFixture::value_type;
+    md_type md[3] = {
+        {gko::dim<2>{6, 2}, {{0, 0, 0}, {0, 1, 1}, {1, 0, 2}, {1, 1, 3}}},
+        {gko::dim<2>{6, 2}, {{2, 0, 4}, {2, 1, 5}, {3, 0, 6}, {3, 1, 7}}},
+        {gko::dim<2>{6, 2}, {{4, 0, 8}, {4, 1, 9}, {5, 0, 10}, {5, 1, 11}}}};
```

Comment:
This is still visible to all ranks. I would maybe also add another test where data is only locally visible.

Reply:
I'm only accessing `md[rank]` so I don't think it is that much different from what you describe.

Reply:
Yes, that is true. But all the data is still visible to all ranks. I feel like it would be more fool-proof to have only local data. But this is only a nit and not something absolutely necessary to add.

Reply:
I will leave it like that. I think this is simpler than splitting everything up per rank.